### ISSUE TITLE
Segfault on large pool

### ISSUE THREAD

**Issue by JesusTheHun** [krakjoe/pthreads#466]
Hello,

I work in big data and I need to process very large rowsets, up to 400 millions. This is a backoffice app so only one script like this will go at a time.
Considering the time required to treat all these rows, I turned myself to ptrheads after I discovered it last week. I have never used MTing so I may have made concept mistakes which I apologize in advance for and humbly ask for correction for it.

I've written some code, which work for a small rowset ( a very few thousand ) but segfault when going larger.

Here is the code : 

```php

/////////////
// CLASSES //
/////////////

class WordCountWorker extends Worker
{
    /**
     * @var callable
     */
    protected $callback;

    public function __construct(callable $callback)
    {
        $this->callback = $callback;
    }
}

class WordCountThread extends Thread
{
    protected $row;
    protected $result;

    public function __construct(array $row)
    {
        $this->row = $row;
    }

    public function run()
    {
        $callback = $this->worker->callback;
        $this->result = $callback($this->row);
    }

    public function getResult()
    {
        return $this->result;
    }
}

//////////////////
// RUNNING CODE //
//////////////////

$pdo = new PDO("mysql:host=localhost;charset=UTF8", 'root', '', array(
    PDO::MYSQL_ATTR_USE_BUFFERED_QUERY => false)
);

$stmt = $pdo->query("SELECT * FROM foo");

$pool = new Pool(4, WordCountWorker::class, [function (array $row) {
    foreach ($row as $attr) {
        $row[$attr] = mb_strtoupper($row[$attr]);
    }
    return $row;
}]);

$stmt->setFetchMode(PDO::FETCH_ASSOC);

while (($arr = $stmt->fetch()) !== false) {
    $pool->submit(new WordCountThread($arr));
}

$m = memory_get_peak_usage(true); // To see if these thread object eat my memory like my GF eats chocolate

$pool->shutdown();
$pool->collect(function ($threaded) {

    // Do stuff

    $threadedResult = (array) $threaded->getResult();
    return count($threadedResult);
});
```

Of course this is a dummy process, to see how it behaves before going further.
One approach I considered is to pass small rowset instead of individual row to worker, but even in this case, it does not scale.
And here I'm stuck.
This runs on Linux, pthreads has been compiled from sources with PHP ZTS 5.6

**Comment by JakeFalcor** [comment]
I don't think collect method does what you think it does.

http://php.net/manual/en/pool.collect.php

**Comment by JesusTheHun** [comment]
I know it doesn't do that, I don't have given time to figure out how to do it right now, since the code I shared with you doesn't currently work ^^

**Comment by bnikola** [comment]
Can you post *exact* message that you get when segfault occurs? Also, I assume that `SELECT * FROM foo` is just an example, seeing you said you have up to 400m rows. I take it PHP process would die before spawning a single thread in that case.

**Comment by JesusTheHun** [comment]
This message comes from apache log :

```
[Fri Jul 31 15:51:50.859848 2015] [core:notice] [pid 8128] AH00052: child pid 8408 exit signal Segmentation fault (11)
```

Yes it is an exemple. A "monolitic" version of the 400M rows process already exists and works very fine. It is just damn slow.

**Comment by JesusTheHun** [comment]
This is what I think is happening about my code :

I fetch the statement, I add each row in the pool which distribute them to the workers' stack. Then each worker deals with his own stack, processing one row at a time.
If I am correct, the only thing that can differ from a small rowset is the size of the stack. The things is when I process 1k rows it works fine, and 5k it doesnt. But 5k rows it nothing in term of memory size. So I'm wondering WHY it happens, if that is what actualy happens.

**Comment by cottton** [comment]
`class WordCountThread extends Thread`
your problem -- a Thread each row :)
try this:
```php
class WordCountWorker extends Worker
{
    public function forEachRow($row)
    { // actually this method could be def in Task so you wouldnt need to def a Worker
        foreach ($row as $attr) {
            $row[$attr] = mb_strtoupper($row[$attr]);
        }
        return $row;
    }
}

class Task extends Collectable
{
    protected
        $row,
        $result;

    public function __construct($row)
    {
        $this->row = $row;
    }
    public function run()
    {
        $this->result = $this->worker->forEachRow($this->row);
        $this->setGarbage();
    }
    public function getResult()
    {
        return $this->result;
    }
}

//////////////////
// RUNNING CODE //
//////////////////

$pdo = new PDO("mysql:host=localhost;charset=UTF8", 'root', '', array(
    PDO::MYSQL_ATTR_USE_BUFFERED_QUERY => false)
);

$stmt = $pdo->query("SELECT * FROM foo");

$pool = new Pool(4, 'WordCountWorker');

$stmt->setFetchMode(PDO::FETCH_ASSOC);

while (($row = $stmt->fetch()) !== false) {
    $pool->submit(new Task($row));
    $pool->collect(
        function($Task){
            if(isset($Task->result)){
                // you got your result here ...

                // Do stuff

                echo $Task->result . PHP_EOL;
                return $Task->isGarbage();
            } 
        }
    );
}
$pool->shutdown(); 
```

**Comment by bnikola** [comment]
So if I got this right, here are the symptoms: you have 400 million rows and apparently the real script works slow. However there's no mention *how* the actual script works, how much data it pulls, what kind of processing it does and whether it updates the database back. Multithreaded PHP will let you use multiple cores at once, but the question is whether you are CPU bound in the first place.

As many people have said multiple times - multithreading, although freakin' awesome, is usually not a proper solution for many tasks. Since you use Apache to run a PHP script (there's no mention whether fastcgi is used or if PHP is embedded via superbly awful mod_php which comes with plethora problems of its own), it's difficult to pinpoint what goes wrong. Segfault means something tried accessing memory that it wasn't supposed to access. 

What you're doing is a task fan-out - the main context fetches the data and then distributes it to workers. The problem in this case is that you already retrieved data within master context and memory is allocated for the master context, since you stick everything in an array. You then share this array between your threads and each thread is altering the array. You're going to have some copying and illegal access happening in there. I suspect that's the reason for the segfault, but I could be mistaken - I'm sure one of the cool and more experienced guys will shed some more light on this. If I'm not mistaken, only an object descending from threads should be shared between threads since others are not thread safe (like an array). 

However, why don't you simply take your original script and have it run within a thread, then spawn more threads? Basically, each thread would establish a DB connection in its `run` method, fetch some data and deal with it. All you have to share between threads is the information which thread should fetch which part of the database (one might fetch first 10k rows, the other next 10k etc., easily done).

But, you said the original script is slow - is that original script dealing with database update after it's done processing the data? If yes, your problem is not the CPU time but I/O.

**Comment by JesusTheHun** [comment]
First, thank you for your insight bnikola.

The original process fetch data from a heavy (not complex, but requiring db calculation) request, therefore I am not a big fan of making the request multiple times.

The process analyse groups of rows, extract some rows, and insert them in a MEMORY table. Actually I have chosen a MEMORY Engine because of the I/O problem you talk about, plus I didn't want to slow the entire process down because it was waiting for MySQL to insert, it stacks a hundred of row, and do the insert. Consequently, I/O is unlikely to be the problem.

After reading some issues, It looks like the author doesnt really like the issue/forum-topic kind. So if that ok with you I would like to recenter on the problem on the segfault, and eventually continue the side discussion about what and how on IRC, freenode #pthreads

Talking about the core issue, I am not storing everything in an array, if by you mean storing the query result rowset in the array ; I would not have enough memory to store the rowset. Data is fetch from the mysql server row by row, and the array I pass to the worker stack (via Pool) is a single row array. So I pass a lot of these, of course.

Writing-time PS ( because yes, it is a long reply ) :
Your response made me realize that indeed, I was working with regular array (which are NOT recommended, as krakjoe repeat everywhere). I switched to Threaded and it now works.
However I need to benchmark the execution time to see if it is really relevant to MT this process.

**Comment by JesusTheHun** [comment]
@cottton Your code suggests a new thread for the whole process, correct ? 
I take note of Pool usage, I will try this out.

I think I indeed, make a mistake by extending Thread. I started by extending Threaded, I don't remember why I switched to Thread.

I've tried your code, it does not work (segfault). I turned FETCH_ASSOC into FETCH_CLASS of threaded, it now works but have the same problem of performance (cpu & memory).

I'm gonna close this issue, since the main problem is fixed : I know what caused the segfault and how to fix it (meaning, Array is evil :D )

I'm opening a new one about how to create a custom pool, because I think the way to fix this is to limit the number of tasks passed to the pool. I'm gonna try to stack a few hundreds of rows into a Threaded (not an array :p), and give it to the pool, so there will be less objects passed to the pool (less memory consumed) and the tasks will iterate over.
Also I need to know when the worker is done about his stack to continue fetching the rows and give it more.

If a guru comes by and can enligh me about the best way to achieve what I am trying to do, he is more than welcome :)

Thanks for your time guys.

**Comment by cottton** [comment]
The memory problem:
https://github.com/krakjoe/pthreads/pull/454
There is a bug. I still dont got it how to "fix" it until its fixed.

BTW: i dont know what you are doin with the rows. 
But perhaps its easier to let the mysql server do the job. 
The server is usually faster and you could use multiply connections to share the work.
Like @bnikola said. 
10 connections â 40.000.000 rows and the work done in the query.

**Comment by bnikola** [comment]
My opinion (and therefore *not* a fact) on this particular problem is that threads are not needed. It doesn't mean I think threading is bad in this scenario, I just think there are solutions that are easier to implement which are also able to utilise all the CPU cores in order to execute the task in parallel. If you drop by freenode/#pthreads, we can discuss it there.

**Comment by mkmanwich** [comment]
This issue doesn't exist on pthreads 2.0.7
Try downgrading and running it again, but change the collectable class to threaded (as it doesn't exist in that version).
you will notice 2 things:
1. If you collect, memory doesn't grow.
2. Destructors are never called, not even once, for workers AND for threaded objects.

Something in the new releases messed things up....

**Comment by JesusTheHun** [comment]
AFAIK my problem was the use of an array instead of a Stackable, since using a Stackable fixed the issue. Am I mistaken ? Was that an unknown-for-me workaround ?

If I downgrade to 2.0.7, you say it will fix some things, but what are the things I won't be able to use beside Collectable ?

**Comment by mkmanwich** [comment]
Actually I meant the other problem you will have which is memory leak.
pthreads 2.0.10 has for pool.
collect doesn't work there.
see #441.
